go/vt/wrangler/keyspace.go

```diff
@@ -337,138 +383,88 @@ func (wr *Wrangler) migrateServedTypesLocked(ctx context.Context, keyspace strin
-			if tc := si.GetTabletControl(servedType); !reverse && tc != nil && tc.DisableQueryService {
-				// This is a forwards migration, and the
-				// destination query service was already in a
-				// disabled state. We need to enable and force
-				// a refresh, otherwise it's possible that both
-				// the source and destination will have query
-				// service disabled at the same time, and
-				// queries would have nowhere to go.
-				if err := si.UpdateDisableQueryService(ctx, servedType, cells, false); err != nil {
-					return err
-				}
-				needToRefreshDestinationTablets = true
-			}
-			if reverse && servedType != topodatapb.TabletType_MASTER {
-				// this is a backwards migration, we need to
-				// disable query service on the destination
-				// shards. (we're not allowed to reverse a
-				// master migration).
-				if err := si.UpdateDisableQueryService(ctx, servedType, cells, true); err != nil {
-					return err
-				}
+			if err := si.UpdateDisableQueryService(ctx, topodatapb.TabletType_MASTER, nil, false); err != nil {
+				return err
 			}
 
-			// for master migration, also disable filtered
-			// replication
-			if servedType == topodatapb.TabletType_MASTER {
-				si.SourceShards = nil
-			}
+			// Final migration. Remove source shards.
+			si.SourceShards = nil
 			return nil
 		})
 		if err != nil {
 			return err
 		}
 	}
-	if needToRefreshDestinationTablets {
-		event.DispatchUpdate(ev, "refreshing destination shard tablets so they restart their query service")
-		for _, si := range destinationShards {
-			wr.RefreshTabletsByShard(ctx, si, []topodatapb.TabletType{servedType}, cells)
-		}
-	}
 
 	// And tell the new shards masters they can now be read-write.
 	// Invoking a remote action will also make the tablet stop filtered
 	// replication.
-	if servedType == topodatapb.TabletType_MASTER {
-		event.DispatchUpdate(ev, "setting destination masters read-write")
-		if err := wr.refreshMasters(ctx, destinationShards); err != nil {
-			return err
-		}
+	event.DispatchUpdate(ev, "setting destination masters read-write")
+	if err := wr.refreshMasters(ctx, destinationShards); err != nil {
+		return err
 	}
 
 	event.DispatchUpdate(ev, "finished")
 	return nil
 }
 
+// updateShardRecords updates the shard records based on 'from' or 'to' direction.
+func (wr *Wrangler) updateShardRecords(ctx context.Context, shards []*topo.ShardInfo, cells []string, servedType topodatapb.TabletType, isFrom bool) (err error) {
+	for i, si := range shards {
+		shards[i], err = wr.ts.UpdateShardFields(ctx, si.Keyspace(), si.ShardName(), func(si *topo.ShardInfo) error {
+			if err := si.UpdateServedTypesMap(servedType, cells, isFrom /* remove */); err != nil {
+				return err
+			}
+			return si.UpdateDisableQueryService(ctx, servedType, cells, isFrom /* disable */)
+		})
+		if err != nil {
+			return err
+		}
+		// For 'to' shards, refresh to make them serve.
+		// The 'from' shards will be refreshed after traffic has migrated.
+		if !isFrom {
```

Comment:
`isFrom` is used here because the direction can be reversed with the `-reverse` flag. You can see in `replicaMigrateServedTypes` that we swap the order based on that flag. Hence the new terminology.